templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/controller/TodoItemsController.java

```diff
@@ -0,0 +1,97 @@
+import com.microsoft.azure.simpletodo.model.TodoState;
+import com.microsoft.azure.simpletodo.repository.TodoItemRepository;
+import com.microsoft.azure.simpletodo.repository.TodoListRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.ResponseEntity;
+import org.springframework.util.CollectionUtils;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.servlet.support.ServletUriComponentsBuilder;
+
+import java.math.BigDecimal;
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.StreamSupport;
+
+@RestController
+@RequiredArgsConstructor
+public class TodoItemsController implements ItemsApi {
+    private final TodoListRepository todoListRepository;
+    private final TodoItemRepository todoItemRepository;
+
+    public ResponseEntity<TodoItem> createItem(String listId, TodoItem todoItem) {
+        final Optional<TodoList> optionalTodoList = todoListRepository.findById(listId);
+        if (optionalTodoList.isPresent()) {
+            todoItem.setListId(listId);
+            final TodoItem savedTodoItem = todoItemRepository.save(todoItem);
+            final URI location = ServletUriComponentsBuilder
+                .fromCurrentRequest()
+                .path("/{id}")
+                .buildAndExpand(savedTodoItem.getId())
+                .toUri();
+            return ResponseEntity.created(location).body(savedTodoItem);
+        } else {
+            return ResponseEntity.notFound().build();
+        }
+    }
+
+    public ResponseEntity<Void> deleteItemById(String listId, String itemId) {
+        return todoItemRepository.findTodoItemByListIdAndId(listId, itemId)
+            .map(i -> todoItemRepository.deleteTodoItemByListIdAndId(i.getListId(), i.getId()))
+            .map(i -> ResponseEntity.noContent().<Void>build())
+            .orElse(ResponseEntity.notFound().build());
+    }
+
+    public ResponseEntity<TodoItem> getItemById(String listId, String itemId) {
+        return todoItemRepository.findTodoItemByListIdAndId(listId, itemId)
+            .map(ResponseEntity::ok)
+            .orElseGet(() -> ResponseEntity.notFound().build());
+    }
+
+    public ResponseEntity<List<TodoItem>> getItemsByListId(String listId, BigDecimal top, BigDecimal skip) {
+        // no need to check nullity of top and skip, because they have default values.
+        return todoListRepository.findById(listId)
+            .map(l -> todoItemRepository.findTodoItemsByTodoList(l.getId(), skip.intValue(), top.intValue()))
+            .map(ResponseEntity::ok)
+            .orElseGet(() -> ResponseEntity.notFound().build());
+    }
+
+    public ResponseEntity<TodoItem> updateItemById(String listId, String itemId, TodoItem todoItem) {
+        // make sure listId and itemId are set into the todoItem, otherwise it will create a new todo item.
+        todoItem.setId(itemId);
+        todoItem.setListId(listId);
```

Comment:
This might have been existing code, but is there a different way we can avoid this commented behavior? Like for example, creating a separate POJO for `UpdateTodoItem` that doesn't include ItemId and ListId, and forcing us to create a `TodoItem` with all the required properties set.

Otherwise, it seems like it's too easy to comment out the two lines here and break the update behavior in an unexpected way.

Reply:
Your approach should work, but will add a lot of such POJOs. Especially when the Application function becomes richer.
Some possible approaches I can think of are
1. Use the existing validation framework to validate the parameters.
2. Use explicit well-defined `insert(TodoItem item)`/`update(TodoItem item)` method instead of a save() method that is compatible with both. And throw exceptions if primary key properties are empty in `update()` method. 
3. or make even more explicit methods like `updateState(@NotNull String listId, @NotNull String itemId, @NotNull TodoState state)`.

But of course, it's not easy to make such mistakes if you are familiar with what `save()` does.

Reply:
I see. We're handling `upserts`. That makes sense then.